src/object-pool.ts

```diff
@@ -1,10 +1,287 @@
-import {ArmorObjectPoolOptions} from './object-pool-options';
-import {ArmorObjectPoolState} from './object-pool-state';
+import ArmorCollection from './collection';
+import ArmorCollectionSelector from './selector';
+import ArmorObjectPoolInstance from './object-pool-instance';
+import ArmorObjectPoolOptions from './object-pool-options';
+import ArmorObjectPoolState from './object-pool-state';
 
-export class ArmorObjectPool<T> {
-	public readonly state: ArmorObjectPoolState;
+export default class ArmorObjectPool<T> implements ArmorCollection<T> {
+	public state: ArmorObjectPoolState<T>;
```

Comment:
Making `ArmorObjectPoolState<T>` readonly is ideal, You probably found that to do so you can't set it after construction and thus it's harder to use clean functions on. The way to achieve this is by returning a state object from your parse functions which I'll explain changes below.